handlers/xdelta_handler.c

```diff
@@ -0,0 +1,263 @@
+    }
+    source->curblkno = source->getblkno;
+    return 0;
+}
+
+static int
+write_to_output_file(handler_data* handle, xd3_stream* stream) {
+    size_t dst_bytes_written = fwrite(stream->next_out, sizeof(stream->next_out[0]), stream->avail_out, handle->dst_file);
+    if (dst_bytes_written != (size_t)stream->avail_out) {
+        ERROR("Failed to write %u bytes to output file %s error %s",
+                stream->avail_out, handle->dst_filename, strerror(errno));
+        return errno;
+    }
+    xd3_consume_output(stream);
+    return 0;
+}
+
+// Implementation
+int xdelta_handler(img_type *img, void __attribute__ ((__unused__)) *data)
+{
+    int xd3ret, ret = 0;
+    xd3_stream stream;
+    xd3_config config;
+    xd3_source source;
+    handler_data handle;
+
+    memset(&handle, 0, sizeof(handle));
+    memset(&stream, 0, sizeof(stream));
+    memset(&source, 0, sizeof(source));
+    memset(&config, 0, sizeof(config));
+
+    // Parse the extra sw-description fields
+    handle.dst_filename = img->device;
+    handle.patch_filename = img->fname;
+    if ((ret = read_sw_description_extras(img, &handle)) != 0) {
+        goto cleanup;
+    }
+    INFO("Starting XDelta diff from %s to %s with %s", handle.src_filename, handle.dst_filename, handle.patch_filename);
+
+    // Open all relevant files
+    if ((ret = open_files(img, &handle)) != 0) {
+        goto cleanup;
+    }
+
+    // Xdelta3 setup
+    set_default_xd3_config(&config, &source);
+    if ((ret = xd3_config_stream(&stream, &config) != 0)) {
+        ERROR("xd3_config_stream failed %d", ret);
+        goto cleanup;
+    }
+    if ((ret = xd3_set_source(&stream, &source) != 0)) {
+        ERROR("xd3_set_source failed %d", ret);
+        goto cleanup;
+    }
+
+    // prepare patch buffer
+    handle.patch_buf_sz = config.winsize;
+    if (!(handle.patch_buf = (void*) malloc(handle.patch_buf_sz))) {
+        ERROR("Patch buffer malloc failed sz %d", handle.patch_buf_sz);
+        ret = -1;
+        goto cleanup;
+    }
+
+    // main processing loop
+    do {
+        // Read patch file data
+        if ((ret = read_from_patch_file(&handle, &stream)) != 0) {
+            goto cleanup;
+        }
+
+process:
+        /* this command does all the magic */
+        xd3ret = xd3_decode_input(&stream);
+        switch (xd3ret) {
+            case XD3_INPUT: /* stream requires more input */            
+                continue;
+
+            case XD3_OUTPUT: /* destimation ( aka "new" ) file segment is ready to be written */
+            {
+                if ((ret = write_to_output_file(&handle, &stream)) != 0) {
+                    goto cleanup;
+                }
+                goto process;
+            }
+
+            case XD3_GETSRCBLK: /* need data from the source file */
+            {
+                if ((ret = read_from_src_file(&handle, &source)) != 0) {
+                    goto cleanup;
+                }
+                goto process;
+            }
+
```

Comment:
Maybe we should say something about these states?

Reply:
I have added comment with a reference to Xdelta programming manual up above.